physical/raft/raft.go

```diff
@@ -902,6 +918,58 @@ func (b *RaftBackend) SetupCluster(ctx context.Context, opts SetupOpts) error {
 	// Close the init channel to signal setup has been completed
 	close(b.raftInitCh)
 
+	reloadConfig := func() {
+		newCfg := raft.ReloadableConfig{
+			TrailingLogs:      raftConfig.TrailingLogs,
+			SnapshotInterval:  raftConfig.SnapshotInterval,
+			SnapshotThreshold: raftConfig.SnapshotThreshold,
+			HeartbeatTimeout:  raftConfig.HeartbeatTimeout / initialTimeoutMultiplier,
+			ElectionTimeout:   raftConfig.ElectionTimeout / initialTimeoutMultiplier,
+		}
+		err := raftObj.ReloadConfig(newCfg)
+		if err != nil {
+			b.logger.Error("failed to reload raft config to set lower timeouts", "error", err)
+		} else {
+			b.logger.Trace("reloaded raft config to set lower timeouts", "config", fmt.Sprintf("%#v", newCfg))
+		}
+	}
+	confFuture := raftObj.GetConfiguration()
+	if err := confFuture.Error(); err != nil {
+		b.logger.Error("failed to read raft configuration", "error", err)
+	}
+	clusterConf := confFuture.Configuration()
+	if initialTimeoutMultiplier != 0 {
+		if len(clusterConf.Servers) == 1 {
+			reloadConfig()
+		} else {
+			go func() {
+				ticker := time.NewTicker(50 * time.Millisecond)
+				// Emulate the random timeout used in Raft lib, to ensure that
+				// if all nodes are brought up simultaneously, they don't all
+				// call for an election at once.
+				extra := time.Duration(rand.Int63()) % raftConfig.HeartbeatTimeout
+				timeout := time.NewTimer(raftConfig.HeartbeatTimeout + extra)
+				for {
+					select {
+					case <-ticker.C:
+						switch raftObj.State() {
+						case raft.Candidate, raft.Leader:
+							b.logger.Trace("triggering raft config reload due to being candidate or leader")
+							reloadConfig()
+							return
+						case raft.Shutdown:
+							return
+						}
+					case <-timeout.C:
+						b.logger.Trace("triggering raft config reload due to initial timeout")
+						reloadConfig()
+						return
+					}
+				}
+			}()
+		}
+	}
+
```

Comment:
I'm a little bit confused about the code inside this if statement. It looks like the election and heartbeat timeouts are only multiplied by the initialTimeoutMultiplier on lines 779 and 780 if the node starting up isn't the leader. `reloadConfig()` undoes that multiplication by dividing those timeouts by the initialTimeoutMultiplier, but it looks like that function is being called on both leader and follower nodes - lines 943 and 958. Wouldn't we only want to reload the config on the follower nodes that had the extra large values? Otherwise, you'd end up dividing normal values by initialTimeoutMultiplier and end up with really small timeouts, no?

Reply:
StartAsLeader is a special option that's used when we know a priori that the current node should be the leader, and in fact the only node.  Example: storage migration, recovery mode, bootstrapping first node of cluster.  In light of that does the code make more sense now?

Reply:
Ahh, yes, that helps a lot, thank you.